lib/run_time.mli

```diff
@@ -0,0 +1,52 @@
+val src : Logs.src
+module Log : Logs.LOG
+
+type timestamps =
+    Queued of float
+  | Running of { ready : float; started : float; }
+  | Finished of { ready : float; started : float option; finished : float; }
+  (** Type to encode the timestamps of a step or a build. All floats represent
+      Unix epoch time - i.e. milliseconds from 1970-01-01 00:00:00 GMT
+
+      The step or build is either:
+      - Queued - the associated timestamp indicates when it was ready to run (enqueued)
+      - Running - with timestamps for when it was ready and when it started
+      - Finished - with timestamps for when it was ready, when it started and when it finished.
+      Note that it may never run -- it may be cancelled or it may timeout *)
+
+type run_time_info =
+    Cached
+  | Queued_for of float
+  | Running of { queued_for : float; ran_for : float; }
+  | Finished of { queued_for : float; ran_for : float option; }
+  (** Type to encode the durations of the various stages of a step or build.
+      All floats represent milliseconds.
+
+      The step or build is either:
+      - Queued - the associated duration represents the number of milliseconds it has been queued for (relative to now)
+      - Running - has a queued_for duration, and a ran_for duration. The latter should be interpreted as the number of
+      milliseconds that it has been running for
+      - Finished - has a queued_for and an optional ran_for duration. The ran_for duration will be None when it did
+      not run. *)
+
+val pp_timestamps : Format.formatter -> timestamps -> unit
+(** ppx_derived Show formatter for timestamps *)
```

Comment:
Should this just be a Fmt function that takes a `timestamp` and leave the IO to the caller. It would be more flexible that way.

Reply:
So this is what results from `[@@deriving show]` and it's easy to use it in tests. I don't expect this to be used further (and would ideally like it to not be part of the interface but I can't use it in tests otherwise).

> Should this just be a Fmt function that takes a timestamp 

Yep - we'd write presenters of this ilk to suit the cli and also the web front-end.